Mathlib/CategoryTheory/Category/Quiv/AsOfFunctor.lean

```diff
@@ -0,0 +1,396 @@
+/-
+Copyright (c) 2025 Robert Maxton. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Robert Maxton
+-/
+
+import Mathlib.CategoryTheory.Category.Quiv
+import Mathlib.CategoryTheory.Functor.Functorial
+import Mathlib.CategoryTheory.Limits.FunctorCategory.EpiMono
+import Mathlib.CategoryTheory.Limits.Types.Filtered
+import Mathlib.Data.Set.Lattice.Image
+import Mathlib.Logic.Small.Set
+import Mathlib.Tactic.ApplyFun
+
+/-!
+  # Quivers as (co-)presheaves on the walking quiver
+
+  In this file, we show that the category `Quiv.{v, u}` is equivalent to the full subcategory of
+  the category **PSh(WPPair)** of presheaves on the walking parallel pair which are
+  `u`-small and locally `v`-small. Here, `WalkingParallelPairᵒᵖ` takes the role of the
+  "walking quiver", the category `1 ⇉ 0`. Every presheaf on the walking quiver picks out:
+
+  * A type for `0`, which we interpret as the type of vertices
+  * A type for `1`, which we interpret as the type of edges (sum of all hom-types)
+  * A function for `left : 1 ⟶ 0`, which we interpret as assigning a source vertex to each edge
+  * A function for `right : 1 ⟶ 0`, which we intepret as assigning a target vertex to each edge
+
+  When `u = v`, `Quiv.{u, u}` is equivalent to the entire category **PSh(WPPair)** and
+  as such is complete and cocomplete, as well as enjoying many other nice properties.
+  Otherwise, we can still generate structure in `Quiv.{v, u}` by showing that relevant
+  properties are closed under being `{u, v}`-small.
+
+  ## Implementation
+
+  We define `asFunctor` and `ofFunctor` as functions first, and then show that
+  they are functorial and give an equivalence of categories. This is primarily
+  so that `asFunctor` can be used with projection notation when the universe
+  levels can be inferred, and because showing the functoriality of `ofFunctor`
+  is easier after defining some API for the base function.
+
+  We define both `asFunctor` and `ofFunctor` with an extra universe level `w`,
+  which allows us to more easily state the equivalence between
+  `SmallAsQuivSubcategory` at size `max v u` and any larger size.
+-/
+
+open CategoryTheory Limits Bicategory Strict UnivLE ObjectProperty
+
+attribute [pp_with_univ] Quiv Quiver
+
+namespace CategoryTheory.Quiv
+open Opposite
+universe w₂ w₁ w' w v₂ v₁ v u₂ u₁ u
+
+scoped instance : Zero (WalkingParallelPairᵒᵖ) := ⟨.op .zero⟩
+scoped instance : One (WalkingParallelPairᵒᵖ) := ⟨.op .one⟩
+
+@[simp← ] lemma WalkingQuiver.zero_def : (0 : WalkingParallelPairᵒᵖ) = .op .zero := rfl
+@[simp← ] lemma WalkingQuiver.one_def : (1 : WalkingParallelPairᵒᵖ) = .op .one := rfl
+
+open WalkingQuiver
+
+/-- Interpret a quiver as a co-presheaf on the walking quiver. -/
+@[simps -fullyApplied, pp_with_univ]
+def asFunctor (Q : Quiv.{v, u}) : WalkingParallelPairᵒᵖ ⥤ Type max w v u where
+  obj
+  | 0 => ULift.{max w v} Q.α
+  | 1 => (s t : ULift.{max w v} Q.α) × (Q.str.Hom s.1 t.1)
```

Comment:
In general, such definitions should be defined first for a type `Q` with a `Quiver` instance rather than on the category `Quiv` directly. Also, it would make sense to introduce a definition for the type of arrows in a quiver, and this should probably we a structure so as to parallel `Arrow` (in case of categories). I am not convinced there is a need for the auxiliary universe `w` (and for the simplicity of formalization, we may even remove all ulifts by restricting to the case of "small" quivers).

Reply:
I repeat: defining things on `Quiv` (or `Cat`) directly is a very bad idea (for example, the parallel definition of the nerve of a category is not defined directly as a functor from `Cat`), and the universe `w` is very unnecesary.